### ISSUE TITLE
Router Engine

### ISSUE THREAD

**Issue by blakeembrey** [pillarjs/router#12]
I need to implement a router similar to this one, but with different path matching semantics. Is there any plan to abstract the core logic into an engine module that other implementations can use?

**Comment by dougwilson** [comment]
So I, for one, want to make the path syntax pluggable, and I'm not sure if that's in line with what you want. Otherwise, I'm always for more splitting of stuff. What part were you think of?

**Comment by scottcorgan** [comment]
I think that making the routes pluggable is a fantastic idea. My specific use case is negated glob-type routes.

**Comment by dougwilson** [comment]
I want URL template-style routes. Not sure if that's what you mean by "glob-style" :)

**Comment by scottcorgan** [comment]
Not sure what you mean by template-style.

By glob-style I mean "!/**/assets/*.js"

**Comment by dougwilson** [comment]
Gotcha. URL templates are RFC 6570 (simple example: `/user/{userId}/friends`).

**Comment by scottcorgan** [comment]
Is it much work to make the paths pluggable?

**Comment by dougwilson** [comment]
I haven't yet tried, but it shouldn't be too hard. I definitely intend to make the paths pluggable (more than just providing a `RegExp`, since `req.params` cannot have named properties with a `RegExp`).

**Comment by scottcorgan** [comment]
Cool. Looking forward to it. Router has become indispensable for us and creating middleware chains.

**Comment by blakeembrey** [comment]
@dougwilson Great, I'll see after and during the development what is similar between the modules and we can factor out. I'm also going to be building it with first-class promise support, so I'll need to see where that fits in with this.

**Comment by dougwilson** [comment]
Cool. So one thing I'm looking forward to is to see if the "io.js" project causes some more movement within Node.js. Because things like potential ES7 `await` make not using Promises really dumb :)

**Comment by scottcorgan** [comment]
So, this is going to be solved by #13?

Any docs on that branch?

**Comment by blakeembrey** [comment]
@scottcorgan It'll fix the engine issue by allowing your own match function to be used. The implementation using `path-to-regexp` in `index.js` is the only documentation right now.

**Comment by scottcorgan** [comment]
What would that look like?

**Comment by blakeembrey** [comment]
Check out https://github.com/blakeembrey/router/blob/router-engine/index.js

You'll need to implement the methods for `.route`, `.use` and `#verb` that will turn the path into a function instead of a regexp. That function just needs to return an object `{ path: '/foo', { params: { id: 'foo } }` or `false` if no match.

**Comment by fourpixels** [comment]
@dougwilson Cheers for moving away from Express and creating PillarJS! That's what open source means!
For the sake of all - **please** rewrite this router as it's driving me crazy! I can't understand how in 2016 there's no way to backtrack those paths and you have to hardcode the links in your views.
What I mean is that paths are now anonymous and because of the middleware style of array-like routes, there's no chance to get back and create link for specific route. For example:

```js
// app.js
app.get('/workspace/:workspaceID/project/:projectID/user/:userID', function(req, res, next) { // ...
```

```html
<!-- view.html; if using any some kind of templating engine -->
<a href="/workspace/{{ workspace.id }}/project/{{ project.id }}/user/{{ user.id }}>link</a>
```

```js
// not to mention creation of a link via pure js:
href = "/workspace/" + workspace.id + "/project/" + project.id + "/user/" + user.id
```

If I want to change `workspace` to `workspaces` I have to edit **all** my views, which will potentially make me die :)

So if you have some spare time - I would love to have this feature, and I'm sure tons of others are needing it too. I would love to contribute, but the more I dive into the Router, the more I'm smashing my head in the wall, as it's so absurd..

**Comment by wesleytodd** [comment]
@fourpixels Sounds like you should open a new issue that can be tagged as `idea`.  Your point seems valid, but could probably be fleshed out in its own issue/thread.

**Comment by fourpixels** [comment]
Yeah, sorry about that. I just saw a router ideas issue, so I didn't want to mess things around before knowing what you guys think. Will open one next with better description, you can delete my posts here :)
